panic when metadata does not exist in an install-config

https://github.com/openshift/hive/blob/54d82744ac5253ff262a8de6061438ef421573d6/pkg/clusterresource/builder.go#L435

When experimenting with clusterpools I found that the hive controllers pod will enter a crashloop state if a user creates an install-config for a clusterpool which does not contain metadata.

Given the docs here: https://github.com/openshift/hive/blob/master/docs/clusterpools.md#install-config-template my expectation is that the controller would safely merge the `metadata.name` even if metadata did not exist.

```
time="2021-06-29T10:17:55.784Z" level=info msg="reconciling cluster pool" clusterPool=test-clusters/atlas-platform-dev controller=clusterpool reconcileID=ln5vzzk5
time="2021-06-29T10:17:55.807Z" level=info msg="Adding new clusters" clusterPool=test-clusters/atlas-platform-dev controller=clusterpool count=1 reconcileID=ln5vzzk5
time="2021-06-29T10:17:55.826Z" level=info msg="Creating new cluster" cluster=atlas-platform-dev-rxvbl clusterPool=test-clusters/atlas-platform-dev controller=clusterpool reconcileID=ln5vzzk5
time="2021-06-29T10:17:55.826Z" level=info msg="reconcile complete" clusterPool=test-clusters/atlas-platform-dev controller=clusterpool elapsedMillis=42 elapsedMillisGT=0 outcome=unspecified reconcileID=ln5vzzk5
E0629 10:17:55.826958       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 871 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2e40060, 0x52ccb90)
	k8s.io/apimachinery@v0.20.0/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	k8s.io/apimachinery@v0.20.0/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2e40060, 0x52ccb90)
	runtime/panic.go:969 +0x175
github.com/openshift/hive/pkg/clusterresource.(*Builder).mergeInstallConfigTemplate(0xc00245a1e0, 0xc00245a5a0, 0x0, 0x0)
	github.com/openshift/hive/pkg/clusterresource/builder.go:435 +0x123
github.com/openshift/hive/pkg/clusterresource.(*Builder).Build(0xc00245a1e0, 0xc00245a1e0, 0xc00149cbc0, 0x1, 0x1, 0xc000ce24d0)
	github.com/openshift/hive/pkg/clusterresource/builder.go:202 +0x168
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).createCluster(0xc00170eba0, 0xc001e01800, 0x39d5ce0, 0xc003aaa360, 0xc000dde800, 0x15fb, 0xc001536380, 0x305, 0x39fdfe0, 0xc001024ee0, ...)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:560 +0x2e5
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).addClusters(0xc00170eba0, 0xc001e01800, 0x1, 0x39fdfe0, 0xc001024ee0, 0x0, 0x0)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:518 +0x76e
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).Reconcile(0xc00170eba0, 0x39bdde0, 0xc000bf6300, 0xc002334ab0, 0xd, 0xc0025f22e0, 0x12, 0xc000bf6300, 0x0, 0x0, ...)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:355 +0x1966
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0016fcc80, 0x39bdd20, 0xc0004478c0, 0x2fc71c0, 0xc001488920)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0016fcc80, 0x39bdd20, 0xc0004478c0, 0x0)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x39bdd20, 0xc0004478c0)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00152af50)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00149df50, 0x395e640, 0xc000c3e870, 0xc000447801, 0xc0001282a0)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00152af50, 0x3b9aca00, 0x0, 0xc0014d7d01, 0xc0001282a0)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x39bdd20, 0xc0004478c0, 0xc00125d020, 0x3b9aca00, 0x0, 0x1)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x39bdd20, 0xc0004478c0, 0xc00125d020, 0x3b9aca00)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x26cc5e3]

goroutine 871 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	k8s.io/apimachinery@v0.20.0/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x2e40060, 0x52ccb90)
	runtime/panic.go:969 +0x175
github.com/openshift/hive/pkg/clusterresource.(*Builder).mergeInstallConfigTemplate(0xc00245a1e0, 0xc00245a5a0, 0x0, 0x0)
	github.com/openshift/hive/pkg/clusterresource/builder.go:435 +0x123
github.com/openshift/hive/pkg/clusterresource.(*Builder).Build(0xc00245a1e0, 0xc00245a1e0, 0xc00149cbc0, 0x1, 0x1, 0xc000ce24d0)
	github.com/openshift/hive/pkg/clusterresource/builder.go:202 +0x168
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).createCluster(0xc00170eba0, 0xc001e01800, 0x39d5ce0, 0xc003aaa360, 0xc000dde800, 0x15fb, 0xc001536380, 0x305, 0x39fdfe0, 0xc001024ee0, ...)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:560 +0x2e5
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).addClusters(0xc00170eba0, 0xc001e01800, 0x1, 0x39fdfe0, 0xc001024ee0, 0x0, 0x0)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:518 +0x76e
github.com/openshift/hive/pkg/controller/clusterpool.(*ReconcileClusterPool).Reconcile(0xc00170eba0, 0x39bdde0, 0xc000bf6300, 0xc002334ab0, 0xd, 0xc0025f22e0, 0x12, 0xc000bf6300, 0x0, 0x0, ...)
	github.com/openshift/hive/pkg/controller/clusterpool/clusterpool_controller.go:355 +0x1966
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0016fcc80, 0x39bdd20, 0xc0004478c0, 0x2fc71c0, 0xc001488920)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0016fcc80, 0x39bdd20, 0xc0004478c0, 0x0)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x39bdd20, 0xc0004478c0)
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00152af50)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00149df50, 0x395e640, 0xc000c3e870, 0xc000447801, 0xc0001282a0)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00152af50, 0x3b9aca00, 0x0, 0xc0014d7d01, 0xc0001282a0)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x39bdd20, 0xc0004478c0, 0xc00125d020, 0x3b9aca00, 0x0, 0x1)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x39bdd20, 0xc0004478c0, 0xc00125d020, 0x3b9aca00)
	k8s.io/apimachinery@v0.20.0/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
```

--- a/pkg/clusterresource/builder.go
+++ b/pkg/clusterresource/builder.go
@@ -432,6 +432,9 @@ func (o *Builder) mergeInstallConfigTemplate() (*corev1.Secret, error) {
 		return nil, fmt.Errorf("Error parsing installconfigtemplate: %s", err.Error())
 	}
 	ic.BaseDomain = o.BaseDomain
+	if ic.MetaData == nil {
+		ic.MetaData = &metav1.ObjectMeta{}
+	}
 	ic.MetaData.Name = o.Name
 
 	d, err := yaml.Marshal(ic)

--- a/pkg/clusterresource/builder_test.go
+++ b/pkg/clusterresource/builder_test.go
@@ -68,6 +68,14 @@ networking:
   serviceNetwork:
   - 172.30.0.0/16
 `
+	fakeMinimalInstallConfigYaml = `apiVersion: v1
+compute:
+- hyperthreading: Enabled
+  name: worker
+platform:
+  aws:
+    region: us-west-2
+`
 )
 
 func createTestBuilder() *Builder {
@@ -246,7 +254,8 @@ func TestBuildClusterResources(t *testing.T) {
 				require.NotNil(t, certSecret)
 				assert.Equal(t, certSecret.Name, cd.Spec.Platform.VSphere.CertificatesSecretRef.Name)
 			},
-		}, {
+		},
+		{
 			name: "merge InstallConfigTemplate",
 			builder: func() *Builder {
 				b := createAWSClusterBuilder()
@@ -265,6 +274,31 @@ func TestBuildClusterResources(t *testing.T) {
 
 				re := strings.NewReplacer("template.domain", baseDomain, "template-cluster-name", clusterName)
 				updatedYaml := re.Replace(fakeInstallConfigYaml)
+
+				assert.YAMLEq(t, updatedYaml, installConfigSecret.StringData["install-config.yaml"])
+			},
+		},
+		{
+			name: "merge MinimalInstallConfigTemplate",
+			builder: func() *Builder {
+				b := createAWSClusterBuilder()
+
+				b.Name = clusterName
+				b.BaseDomain = baseDomain
+				b.Namespace = namespace
+
+				b.InstallConfigTemplate = fakeMinimalInstallConfigYaml
+				yaml.Unmarshal([]byte(fakeMinimalInstallConfigYaml), b.InstallConfigTemplate)
+				return b
+			}(),
+			validate: func(t *testing.T, allObjects []runtime.Object) {
+
+				installConfigSecret := findSecret(allObjects, fmt.Sprintf("%s-install-config", clusterName))
+
+				updatedYaml := fakeMinimalInstallConfigYaml + fmt.Sprintf(`baseDomain: %s
+metadata:
+  creationTimestamp: null
+  name: %s`, baseDomain, clusterName)
 
 				assert.YAMLEq(t, updatedYaml, installConfigSecret.StringData["install-config.yaml"])
 			},

